File: 200-number-of-islands/200-number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int nr = grid.size();
        if(!nr) return false;
        int nc = grid[0].size();
        int num_islands = 0;
        for(int i = 0; i < nr; i++)
        {
            for(int j = 0; j < nc; j++)
            {
                if(grid[i][j] == '1')
                {
                    num_islands++;
                    queue<pair<int,int>> q;
                    grid[i][j] = '0';
                    q.push({i,j});
                    while(!q.empty())
                    {
                        pair<int,int> rc = q.front();
                        q.pop();
                        int row = rc.first;
                        int col = rc.second;
                        if(row - 1 >= 0 && grid[row - 1][col] == '1' )
                        {
                            q.push({row-1, col});
                            grid[row - 1][col] = '0';
                        }
                        if(row + 1 < nr && grid[row + 1][col] == '1' )
                        {
                            q.push({row+1, col});
                            grid[row + 1][col] = '0';
                        }
                        if(col - 1 >= 0 && grid[row][col - 1] == '1' )
                        {
                            q.push({row, col - 1});
                            grid[row][col -1] = '0';
                        }
                        if(col + 1 < nc && grid[row][col + 1] == '1' )
                        {
                            q.push({row, col + 1});
                            grid[row][col + 1] = '0';
                        }
                    }
                    
                }
            }
        }
        return num_islands;
    }
};

File: 200-number-of-islands/200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        
        if not grid or not grid[0]:
            return 0
        numIsland = 0
        
        rows, cols = len(grid), len(grid[0])
        seen = set()
        def dfs(r , c):
            if r not in range(rows) or c not in range(cols) or grid[r][c] == "0" or (r,c) in seen:
                return
            seen.add((r,c))
            directions = [[0,1],[0,-1],[1,0],[-1,0]]
            
            for dr, dc in directions:
                dfs(r + dr, c + dc)
                
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1" and (r,c) not in seen:
                    numIsland += 1
                    dfs(r,c)
        return numIsland
